### PR TITLE
Added feature to ignore indexes starting with dot during shard limit validation

### LINKED ISSUE #4675: Add feature to ignore indexes starting with dot on shard limit validation.
**Is your feature request related to a problem? Please describe.**
Today on the ShardLimitValidator as mention here [code](https://github.com/opensearch-project/OpenSearch/blob/7b75fb425f9df2b96049dac8c7dc13273f5026fd/server/src/main/java/org/opensearch/indices/ShardLimitValidator.java#L98) only system indexes are ignored from shard limit validation however other plugins which do not inherit system index properties [may be security or hidden indexes] are not excluded. So a feature is required to ignore these indexes from the shard validation phase. 

**Describe the solution you'd like**
The solution until all plugin follow a standard method of inheriting the system indices we can exempt indexes starting with a '.' character. As most of the plugins create an index which starts with '.' so they will be ignored during the shard limit validation phase. 

Further, We can use a cluster setting as a feature flag for a gated mechanism where if the feature flag is enabled we ignore the indexes during shard limit validation otherwise handle those indexes as normal indexes. 

**Describe alternatives you've considered**
Move all plugins to follow system indices instead of dot indices as discussed here #3942

### PR DESCRIPTION
**Pull request by nishchay21**
Signed-off-by: Nishchay Malhotra <nishcha@amazon.com>

### Description
The Current PR helps in solving the issue #4675.

Currently, all plugins do not follow a standard method of inheriting the system indices and are not ignored during the shard limit validation phase. The goal of this PR is to exempt indexes starting with a '.' character to be ignored during this shard limit validation phase. 

Further, this PR also adds a cluster setting "cluster.ignore_dot_indexes" which works as a gated mechanism for the feature to be enabled or disabled. This is a dynamic cluster setting which takes in a boolean value. If the setting is set to true the indexes starting with the '.' character are ignored otherwise considered like other normal indexes. Adding to this the default value of this is added to be true so that the indexes created with '.' are ignored if this setting is not explicitly set.

Created a new pull request https://github.com/opensearch-project/OpenSearch/pull/4695 as this one had some issues

### Issues Resolved
#4675 [Add feature to ignore indexes starting with dot on shard limit validation]

### Check List
- [x] New functionality includes testing.
- [x] All tests pass
- [x] New functionality has been documented.
- [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -114,6 +114,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
   - Addition of Doc values on the GeoShape Field
   - Addition of GeoShape ValueSource level code interfaces for accessing the DocValues.
   - Addition of Missing Value feature in the GeoShape Aggregations.
+- Added feature to ignore indexes starting with dot during shard limit validation.([#4674](https://github.com/opensearch-project/OpenSearch/pull/4674))
```
Comment (Bukhtawar) [inline CHANGELOG.md:117]: nit: space

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java [hunk 2]
```diff
@@ -64,12 +65,23 @@ public class ShardLimitValidator {
...
+
     protected final AtomicInteger shardLimitPerNode = new AtomicInteger();
+    protected final AtomicBoolean ignoreDotIndexes = new AtomicBoolean();
     private final SystemIndices systemIndices;
 
     public ShardLimitValidator(final Settings settings, ClusterService clusterService, SystemIndices systemIndices) {
         this.shardLimitPerNode.set(SETTING_CLUSTER_MAX_SHARDS_PER_NODE.get(settings));
+        this.ignoreDotIndexes.set(SETTING_CLUSTER_IGNORE_DOT_INDEXES.get(settings));
         clusterService.getClusterSettings().addSettingsUpdateConsumer(SETTING_CLUSTER_MAX_SHARDS_PER_NODE, this::setShardLimitPerNode);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(SETTING_CLUSTER_IGNORE_DOT_INDEXES, this::setIgnoreDotIndexes);
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:84]: can we call it hidden indices?

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java [hunk 3]
```diff
@@ -64,12 +65,23 @@ public class ShardLimitValidator {
         Setting.Property.Dynamic,
         Setting.Property.NodeScope
     );
+
+    public static final Setting<Boolean> SETTING_CLUSTER_IGNORE_DOT_INDEXES = Setting.boolSetting(
+        "cluster.ignore_dot_indexes",
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:70]: hidden?

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java [hunk 4]
```diff
@@ -99,6 +123,14 @@ public void validateShardLimit(final String indexName, final Settings settings,
...
         }
 
+        /*
+        Validates if cluster.ignore_dot_indexes is set to true.
+        If so then it does not validate any index which starts with '.' character.
+        */
+        if (getIgnoreDotIndexes() && indexName.charAt(0) == '.') {
+            return;
+        }
+
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:133]: lets move this to a common logic

**Review by Bukhtawar — COMMENTED** [review]
Thanks @nishchay21 for the changes

File: CHANGELOG.md (hunk 1 above)
Comment (nishchay21) [reply CHANGELOG.md:117]: The above comments to this one are sub comments however the space is same as how parent comments are defined.

File: CHANGELOG.md [hunk 5]
```diff
@@ -128,6 +128,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
```
Comment (Bukhtawar) [inline CHANGELOG.md:131]: nit: indentation?

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java [hunk 6]
```diff
@@ -64,12 +65,23 @@ public class ShardLimitValidator {
...
+    public static final Setting<Boolean> SETTING_CLUSTER_IGNORE_HIDDEN_INDEXES = Setting.boolSetting(
+        "cluster.ignore_hidden_indexes",
+        true,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
     protected final AtomicInteger shardLimitPerNode = new AtomicInteger();
+    protected final AtomicBoolean ignoreHiddenIndexes = new AtomicBoolean();
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:77]: why protected? can we just use a `volatile` boolean

File: CHANGELOG.md (hunk 5 above)
Comment (nishchay21) [reply CHANGELOG.md:131]: The above comments to this one are sub comments however the space is same as how parent comments are defined.

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java [hunk 7]
```diff
@@ -140,6 +173,10 @@ private static int getTotalShardCount(ClusterState state, Index index) {
         return indexMetadata.getNumberOfShards() * (1 + indexMetadata.getNumberOfReplicas());
     }
 
+    private boolean validateHiddenIndex(String indexName) {
+        return getIgnoreHiddenIndexes() && indexName.charAt(0) == '.';
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:177]: Nit - Do we have any definition or disclosure on what happens when an index is created with first character being `.`. If OpenSearch users are creating indices starting with dot, then possible we miss the shard limit validation?

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java (hunk 6 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:77]: +1 -- this can be private and volatile.

**Review by ashking94 — COMMENTED** [review]
LGTM, pls address the comments.

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java (hunk 7 above)
Comment (nishchay21) [reply server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:177]: Usually indices starting with dot  are created by plugin itself however if users create with dot then yes those indexes will not be validated in the shard limit validation.

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java [hunk 8]
```diff
@@ -121,8 +149,13 @@ public void validateShardLimit(final String indexName, final Settings settings,
      */
     public void validateShardLimit(ClusterState currentState, Index[] indicesToOpen) {
         int shardsToOpen = Arrays.stream(indicesToOpen)
-            // Validate shard limit only for non system indices as it is not hard limit anyways
+            /*
+            Validate shard limit only for non system indices as it is not hard limit anyways.
+            Further also validates if the cluster.ignore_hidden_indexes is set to true.
+            If so then it does not validate any index which starts with '.'.
```
[the remark this answers is not in the document]
Comment (nishchay21) [reply server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:155]: ACK will add the same

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java [hunk 9]
```diff
@@ -121,8 +149,13 @@ public void validateShardLimit(final String indexName, final Settings settings,
...
+            */
```
[the remark this answers is not in the document]
Comment (nishchay21) [reply server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:156]: ACK will add the same

File: server/src/main/java/org/opensearch/indices/ShardLimitValidator.java (hunk 6 above)
Comment (nishchay21) [reply server/src/main/java/org/opensearch/indices/ShardLimitValidator.java:77]: ACK.

**Comment by nishchay21** [comment]
gradle check

**Pull request closed without merging**